File: plugin/hooks/hosts/opencode.py
"""OpenCode, as a `Host` record.

Every value here was measured against opencode 1.18.20 on 2026-08-31 with a throwaway
plugin, not read off a docs page. Where a value could not be measured it is the spelling
that means "absent" rather than a hopeful default, because an absent key is the only one
that cannot be mistaken for a working one.

**This host does not run shell hooks at all**, which is the whole reason `js/` exists.
OpenCode's plugin API is in-process JavaScript: a module exporting async functions that
mutate typed output objects. There is no stdin, no stdout, and no `additionalContext`
anywhere in it. `js/opencode.mjs` is the module OpenCode loads; it frames each call as
the JSON payload `run.py` already reads, spawns the same Python, and applies the flat
reply by mutating OpenCode's objects. So the four bodies port unchanged and the host
difference stays data, exactly as it does for a shell host.

Three measurements shaped the mapping and are worth keeping next to it:

* A hook that is `await`ed **stalls the turn** -- 8.016s measured between a handler's
  entry and exit with an 8s sleep in it. Not awaiting exits in 1ms and the detached work
  still finished 8.1s later, because an OpenCode plugin lives in the persistent server
  process. That is why `capture` is mapped to an event the shim does not await, and why
  `supports_async` is True here where Codex, whose async hooks die with the process,
  would have to say otherwise.
* A part pushed into `output.parts` **must carry `id`, `sessionID` and `messageID`**.
  Pushing `{type, text}` alone fails server-side schema validation and kills the entire
  turn with an opaque `UnknownError: Unexpected server error`; the real cause appears only
  in `~/.local/share/opencode/log/opencode.log` as `invalid user part before save`. The
  shim builds the full part; nothing here can express that, which is why the shim is code
  and not another record field.
* Registration is by **directory convention**. `.opencode/plugin/*.js` is auto-loaded
  whether or not the config's `plugin` array names it, so emptying that array does not
  deregister anything -- only removing the file does.
"""

from __future__ import annotations

from core.host import (CODEX_CLI, OPENCODE_CLI, ApproveSpec, Host,
                       TranscriptSpec)

HOST = Host(
    id="opencode",
    #: No plugin-root environment variable exists on this host. The module resolves its
    #: own directory with `import.meta.dirname` and passes the interpreter an absolute
    #: path, so nothing here needs expanding and an empty tuple is the honest value.
    plugin_root_env=(),
    #: Canonical hook -> the OpenCode hook the shim attaches it to.
    #:
    #: `session_start` and `recall` share one OpenCode hook because OpenCode has no
    #: session-start event that can inject: every hook that fires once per session is
    #: `void` and returns nothing. `chat.message` is the only per-message hook with a
    #: mutable output, so the shim runs `session_start` on the first message of a session
    #: and `recall` on every one -- which is the same cadence Claude Code produces from
    #: two separate events, and is possible only because the plugin process persists.
    #:
    #: `capture` rides the generic event stream filtered to `session.idle`. There is no
    #: turn-end hook in the plugin interface at all; `session.idle` is the event that
    #: means the turn stopped.
    events={
        "session_start": "chat.message",
        "recall": "chat.message",
        "capture": "session.idle",
        "approve": "permission.ask",
    },
    #: The shim frames the payload, so these keys are ours rather than the host's. They
    #: are spelled exactly as Claude Code spells them on purpose: one wire vocabulary
    #: between `js/` and `run.py` means a reader comparing two hosts' records is looking
    #: at a real difference when the values differ, not at a shim's private taste.
    fields={
        "session": ("session_id",),
        "cwd": ("cwd",),
        "prompt": ("prompt",),
        "transcript_path": ("transcript_path",),
        "tool_name": ("tool_name",),
    },
    #: Read by `js/shim.mjs`, never by OpenCode. See `envelope._render_flat`.
    envelope="flat",
    context_key="additionalContext",
    #: OpenCode shows a plugin no operator-visible line. Its hooks return data structures
    #: rather than printing, and nothing in the plugin interface renders a status string
    #: to the person at the terminal. Empty here makes the renderer drop that half of a
    #: reply instead of addressing a key nobody reads -- so on this host the account of
    #: what a hook did is its log file, and only its log file.
    status_key="",
    #: Uncapped. Injected parts are ordinary message content here rather than a separate
    #: context channel, so the only budget is `recall.BUDGET`'s.
    context_token_cap=0,
    #: Measured: a detached promise outlives the turn. See the module docstring.
    supports_async=True,
    #: The shim simply does not await capture, so nothing here has to fork.
    detach_capture=False,
    #: Injected parts are ordinary message content, with no cap to declare.
    context_limit_key=0,
    #: Enforced by the shim, which kills the child, because OpenCode publishes no hook
    #: timeout of its own -- an unbounded handler would simply hold the turn open.
    timeouts={"session_start": 20, "recall": 10, "capture": 120, "approve": 5},
    client_configs=("~/.config/opencode/opencode.json",
                    "~/.config/opencode/opencode.jsonc"),
    config_format="json",
    #: OpenCode hands a hook no transcript path, so the shim materialises one: it reads
    #: the session's messages through the plugin's own `client` and writes them as the
    #: same JSONL entries `lib.transcript` already parses. `TranscriptSpec` is present
    #: rather than None because capture CAN run here -- the file just has to be made.
    transcript=TranscriptSpec(format="jsonl"),
    #: The tools whose use is evidence a turn did something, in OpenCode's spelling.
    tools=frozenset({"edit", "write", "bash", "patch"}),
    #: OpenCode wraps no markup of its own around user text. Ours are host-neutral and
    #: live in `lib.transcript.RECALL_MARKERS`.
    noise=(),
    #: A leading slash is a command here as it is everywhere; `!` and `#` are Claude
    #: Code's own and are not OpenCode's, so they are not carried across.
    skip_prefixes=("/",),
    #: OpenCode has no background-task or cross-session prompt wrapper to skip.
    machine_prompt_prefixes=(),
    #: `session.idle` carries no re-entry flag: it fires when the turn stops, and a
    #: continuation is a new turn rather than a flagged repeat of this one.
    reentry_field="",
    #: `permission.ask` gives a mutable `output.status` of "ask" | "deny" | "allow" --
    #: a cleaner permission surface than any other host measured, because the verdict is
    #: one field rather than a decision plus a reason. There is nowhere to put a reason,
    #: so `reason_key` addresses a key the shim drops.
    #:
    #: **The only mapping here not confirmed by a receipt.** No permission prompt fired
    #: during the spike, so which input field carries the tool name is read off the type
    #: definitions rather than measured. `js/opencode.mjs` logs the keys it is handed on
    #: the first call so the shape becomes known from a real invocation, and if the guess
    #: is wrong the match simply misses: nothing is auto-approved and the user is asked,
    #: which is what happens today anyway.
    approve=ApproveSpec(
        matcher="memvara",
        separators=("__", "_"),
        decision_key="status",
        reason_key="reason",
        allow="allow",
    ),
    #: OpenCode's own CLI, so a turn is mined by the model this user already configured
    #: rather than by a different vendor's product. `claude -p` remains the second rung.
    extractor=OPENCODE_CLI,
    #: Distinct from Claude Code's so `memory_why` can tell a user which client wrote a
    #: claim. Frozen from here on for the same reason the Claude one is: it is stored.
    extractor_label="opencode-hook",
    description=(
        "Memvara for OpenCode: recall on every message, capture when the session goes "
        "idle. OpenCode loads JavaScript rather than running shell hooks, so index.mjs "
        "frames each call and spawns run.py, which binds hosts/opencode.py before "
        "dispatching. Capture is not awaited, so a 12-14s extraction never holds the "
        "turn open; nothing this plugin does is visible in the OpenCode UI, so its "
        "account is ~/.memvara/.hooks/capture.log."
    ),
)

File: plugin/hooks/js/opencode.mjs
/**
 * The module OpenCode loads. Maps its in-process hooks onto the four canonical ones.
 *
 * OpenCode's plugin API is unlike every shell host: handlers receive typed objects and
 * inject by mutating them. So this file is the whole of the translation, and it is kept
 * as thin as the host allows -- the memory work itself is the same Python that runs
 * everywhere else, reached through `shim.mjs`.
 *
 * Three behaviours here are measurements rather than choices, each recorded in
 * `hosts/opencode.py` with the numbers:
 *
 * 1. A part pushed into `output.parts` MUST carry `id`, `sessionID` and `messageID`.
 *    Pushing `{type, text}` alone fails schema validation server-side and kills the whole
 *    turn with an opaque `UnknownError`, whose real cause appears only in opencode's own
 *    log as `invalid user part before save`.
 * 2. `capture` is never awaited. Awaiting holds the turn open for the full extraction.
 * 3. `session_start` runs on the first message of each session, because OpenCode has no
 *    session-start hook that can inject -- every once-per-session hook it offers is void.
 */

import { fileURLToPath } from "node:url"

import { note, runHook, runHookDetached } from "./shim.mjs"

const HOST = "opencode"
//: `fileURLToPath`, not `new URL(...).pathname`. The latter yields `/C:/Users/...` on
//: Windows -- a string that looks like a path, joins like a path, and resolves to
//: nothing, so every hook would degrade to "no run.py" on that platform and only there.
const HOOKS_DIR = fileURLToPath(new URL("..", import.meta.url))

/**
 * Sessions whose `session_start` has already run.
 *
 * In memory rather than on disk, and correct only because an OpenCode plugin is loaded
 * once into a server process that outlives the turn -- the same property that makes
 * detached capture work. If that ever stops being true this degrades to running
 * `session_start` more often, never to running it never -- and the entry is released
 * again when the hook did not run, so a transient failure degrades the same way rather
 * than suppressing the standing block for the rest of the session.
 */
const started = new Set()

/** One line, once per process, recording what `permission.ask` actually hands a hook. */
let askShapeLogged = false

const TIMEOUTS = { session_start: 20000, recall: 10000, capture: 120000,
                   approve: 5000, transcript: 15000 }

export const MemvaraPlugin = async ({ client, directory, worktree }) => {
  note("hooks", `opencode plugin loaded dir=${HOOKS_DIR}`)

  /** Materialise a transcript OpenCode never hands us, in the shape `lib.transcript` reads. */
  const writeTranscript = async (sessionID) => {
    try {
      // Bounded, like every other call out of this file. `runHook` enforces a timeout
      // because the host publishes none; this call had none at all, and it sits BEFORE
      // the detached capture -- so a wedged server would hold the event handler open
      // forever and the "capture cannot stall anything" property would not cover the
      // transcript step capture depends on.
      const res = await Promise.race([
        client.session.messages({ path: { id: sessionID } }),
        new Promise((_, reject) =>
          setTimeout(() => reject(new Error("session.messages timed out")),
                     TIMEOUTS.transcript)),
      ])
      const rows = res?.data ?? res ?? []
      const lines = []
      for (const row of rows) {
        const info = row?.info ?? row
        const role = info?.role
        if (role !== "user" && role !== "assistant") continue
        const content = (row?.parts ?? [])
          .filter((p) => p?.type === "text" && p.text)
          .map((p) => ({ type: "text", text: p.text }))
        if (content.length) lines.push(JSON.stringify({ type: role, message: { content } }))
      }
      if (!lines.length) return ""
      const fs = await import("node:fs")
      const os = await import("node:os")
      const path = await import("node:path")
      const dir = path.join(os.tmpdir(), "memvara-opencode")
      fs.mkdirSync(dir, { recursive: true })
      // One file per session, rewritten each turn, so this never grows within a session.
      // Across sessions it would grow without bound -- a machine that has run OpenCode
      // for a month would hold a file per session it ever opened, each the size of a
      // whole conversation. Pruned by age on write rather than deleted after capture,
      // because capture is detached and deleting under a running child is a race.
      const cutoff = Date.now() - 24 * 60 * 60 * 1000
      for (const name of fs.readdirSync(dir)) {
        try {
          const full = path.join(dir, name)
          if (fs.statSync(full).mtimeMs < cutoff) fs.unlinkSync(full)
        } catch { /* another turn pruned it first */ }
      }
      const file = path.join(dir, `${sessionID}.jsonl`)
      fs.writeFileSync(file, lines.join("\n") + "\n")
      return file
    } catch (err) {
      note("hooks", `transcript unavailable session=${sessionID} ${String(err)}`)
      return ""
    }
  }

  return {
    "chat.message": async (input, output) => {
      const sessionID = input?.sessionID ?? ""
      const messageID = output?.message?.id ?? input?.messageID ?? ""
      const prompt = (output?.parts ?? [])
        .filter((p) => p?.type === "text" && p.text)
        .map((p) => p.text)
        .join("\n")

      // Point 1 in this file's header is an invariant about the part we push, so it has
      // to be checked before the work rather than asserted in prose and then defaulted
      // away. Injecting is the whole point, but a part built from ids we do not have is
      // the one thing measured to take the entire turn down with an opaque error, and
      // one missed recall costs a turn's memories rather than the turn.
      if (!sessionID || !messageID) {
        note("hooks", `skipped=chat.message has no ids session=${!!sessionID} ` +
                      `message=${!!messageID}`)
        return
      }

      const payload = { session_id: sessionID, cwd: directory ?? worktree ?? "", prompt }
      const blocks = []

      if (!started.has(sessionID)) {
        // Claimed BEFORE the await so two messages racing into the same new session
        // cannot both run it, and released again if it did not run -- `null` from the
        // shim means exactly that, as opposed to `{}` for "ran, nothing stored yet".
        // Without the release a single 20s timeout on the first message of a session
        // would suppress the standing block for the rest of that session, silently.
        started.add(sessionID)
        const reply = await runHook({
          hooksDir: HOOKS_DIR, hook: "session_start", host: HOST,
          payload, timeoutMs: TIMEOUTS.session_start,
        })
        if (reply === null) started.delete(sessionID)
        else if (reply.additionalContext) blocks.push(reply.additionalContext)
      }

      const reply = await runHook({
        hooksDir: HOOKS_DIR, hook: "recall", host: HOST,
        payload, timeoutMs: TIMEOUTS.recall,
      })
      if (reply?.additionalContext) blocks.push(reply.additionalContext)
      if (!blocks.length) return

      // Every required key, for the reason in this file's header.
      output.parts.push({
        id: `prt_memvara_${Date.now().toString(36)}`,
        sessionID,
        messageID,
        type: "text",
        text: blocks.join("\n\n"),
      })
    },

    "permission.ask": async (input, output) => {
      // The one hook here whose input shape was NOT measured: a permission prompt never
      // fired during the spike, so which field carries the tool name is inferred from
      // the type definitions rather than from a receipt. The keys are logged once so the
      // first real invocation says what actually arrives, and the failure mode if the
      // guess is wrong is benign -- the match misses, nothing is auto-approved, and the
      // user is asked exactly as they are today.
      if (!askShapeLogged) {
        askShapeLogged = true
        note("hooks", `permission.ask keys=${Object.keys(input ?? {}).join(",")}`)
      }
      const tool = input?.type ?? input?.permission ?? input?.title ?? ""
      const reply = await runHook({
        hooksDir: HOOKS_DIR, hook: "approve", host: HOST,
        payload: { session_id: input?.sessionID ?? "", tool_name: String(tool) },
        timeoutMs: TIMEOUTS.approve,
      })
      // Only ever widens to "allow". A hook that could deny would be able to block a
      // tool call the user asked for, which is not what auto-approving reads is for.
      if (reply?.status === "allow") output.status = "allow"
    },

    event: async ({ event }) => {
      if (event?.type !== "session.idle") return
      const sessionID = event?.properties?.sessionID ?? event?.properties?.sessionId ?? ""
      if (!sessionID) return
      const transcript = await writeTranscript(sessionID)
      if (!transcript) return
      // Not awaited: see shim.runHookDetached.
      runHookDetached({
        hooksDir: HOOKS_DIR, hook: "capture", host: HOST,
        payload: { session_id: sessionID, cwd: directory ?? worktree ?? "",
                   transcript_path: transcript },
        timeoutMs: TIMEOUTS.capture,
      })
    },
  }
}

export default MemvaraPlugin
